### ISSUE TITLE
Next steps for `Url` like types

### ISSUE THREAD

**Issue by sydney-runkle** [pydantic/pydantic#10783]
In https://github.com/pydantic/pydantic/pull/10766 and https://github.com/pydantic/pydantic/pull/10662, we made the switch for url like types to be proper Python classes, which helps to clean up behavior in terms of `isinstance` checks and type checking for various constraints (ex - if `host_required` is `True` for a given url type, `host` is typed as `str` rather than `str | None`).

There's still more to be done to clean up this implementation, and more to be done on the URL front in general. The long term goal is to have url structures sort of mimic the proposal outlined here: https://github.com/pydantic/pydantic/issues/10032.

One unique thing about url types is that we validate on `__init__`, sort of like we do for `BaseModel` or `pydantic_dataclass` types.

There are a few specific improvements that I think we could make in the short term:
* We could use a `SchemaValidator` instead of a `TypeAdapter` internally, which might simplify "schema build" time for said types.
* We should simplify the `PyUrl` and `PyMultiHost` url types if possible, as they're no longer meant to be public - perhaps a more simple struct could work well?

One idea we tossed around for a bit was adding a `cls` attribute to the `url_schema` construct in `pydantic-core`, so we could construct the appropriate type there. However, this doesn't make a lot of sense, given that we're currently just storing a `_url` attribute on the more constrained url classes. So, if we were to move to this `cls` construction approach, I don't think the more `RustModel` like approach would make sense here.

There are many change requests and issues associated with urls that parallel this issue, a few of which are seen here:
* https://github.com/pydantic/pydantic/issues/7186
* https://github.com/pydantic/pydantic/issues/8244
* https://github.com/pydantic/pydantic/issues/9794
* https://github.com/pydantic/pydantic/issues/8061

**Comment by emaheuxPEREN** [comment]
I upgraded to the 2.10 beta and I was surprised by the new behavior of urls (seems like new Url classes miss smart equality):
```python
assert AnyHttpUrl('https://pydantic/') == AnyHttpUrl('https://pydantic/')  # AssertionError
```

```
             pydantic version: 2.10.0b1
        pydantic-core version: 2.26.0
```

**Comment by sydney-runkle** [comment]
@emaheuxPEREN,

Thanks for reporting. Will fix this for the official v2.10 release :).

**Comment by sanderr** [comment]
I noticed that #10662 introduced a regression (for lack of a better word) versus 2.9.2, but I admit it's somewhat debatable. So I thought I'd start with a question here rather than to create a dedicated ticket.

The root of the url hierarchy (at least conceptually, because there was no inheritance involved) used to be `AnyUrl`, which was an alias to `pydantic_core._pydantic_core.Url`. Other url types were annotated versions of the same type. Since #10662 this is no longer the case, instead `pydantic.networks._BaseUrl` is now the root of the hierarchy.

I have two concerns regarding this:
1. the docstring that was added to `AnyUrl` in the above-linked PR declares that it is the "Base type for all URLs". This doesn't seem to be accurate. Is the docstring wrong, or is the implementation supposed to match this description?
2. `_BaseUrl` is not part of the public / stable API. So it doesn't seem to be possible anymore to express "any url *type*, don't care which" to represent the actual object (i.e. outside of pydantic parsing).

The case where I'm observing this change of behavior is quite niche and deals with an `issubclass(..., AnyUrl)` check, which might be considered outside of pydantic's contract anyway, but below is an example that I think might have some relevance, though I could also see how you'd consider it to be within the bounds of pydantic validation's invariants.

```python
import pydantic


http_url: pydantic.AnyHttpUrl = pydantic.AnyHttpUrl("https://localhost")
assert isinstance(http_url, pydantic.AnyUrl)  # fails on 2.10.0b2, there's no way to achieve this anymore, this is the case I personally care about most

# validate a Python object that is already a valid url
url: pydantic.AnyUrl = pydantic.TypeAdapter(pydantic.AnyUrl).validate_python(http_url)
assert url is http_url  # fails on 2.10.0b2 => url is no longer an AnyHttpUrl, we lose dynamic type information
```

@sydney-runkle how do you view this change of behavior? Is it possible the docstring is inaccurate? Is there any chance that the top of the hierarchy could be exposed as part of the public API (again)?

**Comment by sydney-runkle** [comment]
> the docstring that was added to AnyUrl in the above-linked PR declares that it is the "Base type for all URLs". This doesn't seem to be accurate. Is the docstring wrong, or is the implementation supposed to match this description?

Good point - the docstring is wrong here. Will fix before v2.10.0 next week.

> _BaseUrl is not part of the public / stable API. So it doesn't seem to be possible anymore to express "any url type, don't care which" to represent the actual object (i.e. outside of pydantic parsing).

I think you can just use `AnyUrl` here?

Regarding the case above:

```py
import pydantic


http_url: pydantic.AnyHttpUrl = pydantic.AnyHttpUrl("https://localhost")
assert isinstance(http_url, pydantic.AnyUrl)
```

Indeed, this change is breaking, but we consider this an improvement in behavior. The following used to fail:

```py
assert isinstance(http_url, pydantic. AnyHttpUrl)
```

Which seems **more** problematic than `assert isinstance(http_url, pydantic.AnyUrl)` failing.

```py
# validate a Python object that is already a valid url
url: pydantic.AnyUrl = pydantic.TypeAdapter(pydantic.AnyUrl).validate_python(http_url)
assert url is http_url
```

I understand the concern here... hmm... maybe all types should still subclass from `AnyUrl`, which is built atop `_BaseUrl`.

Let me what with the team about this.

**Comment by sydney-runkle** [comment]
Thanks so much for bringing this to our attention. Your detailed explanation was very helpful!

You can disregard my justifications above. I agree, we should preserve the `AnyUrl` subclassing behavior.
With [this](https://github.com/pydantic/pydantic/pull/10856) PR:

* All of your assertions in the test code above pass :)
* `assert isinstance(http_url, pydantic. AnyHttpUrl)` also passes
* The docstring on `AnyUrl` is now accurate :)

We'll get this across the line before v2.10.0 next week! Thanks for trying out the beta.

**Comment by sanderr** [comment]
Great, thanks for the swift resolution!

**Comment by Lando-L** [comment]
There also seems to be a bug when serializing and deserializing `AnyUrl`s.

```
from pydantic import AnyUrl, BaseModel

class Test(BaseModel):
    url: AnyUrl

expected = Test(url=AnyUrl("http://example.com"))
json = expected.model_dump_json()
parsed = Test.model_validate_json(json)

expected == parsed # False

expected.url._url # AnyUrl('http://test.example.com/')
parsed.url._url # Url('http://test.example.com/')
```

**Comment by sydney-runkle** [comment]
This is fixed in v2.10.1 :)

**Comment by sanderr** [comment]
@sydney-runkle with this fix, what is the purpose of `pydantic_core.Url`? Is it realistically still expected anywhere in a pydantic model, or will it always be `pydantic.AnyUrl` (and children)?

Motivation for my question: we have a custom json serializer, somewhat similar in concept to `pydantic_core.to_json()` but more tailored to our use case (with some legacy backwards compatibility cases thrown in). We used to ensure url compatibility with an `isinstance(o, pydantic_core.Url)`, which doesn't match anymore with pydantic `2.10`. It's unclear to me whether I should update to `pydantic.AnyUrl`, or to cover both cases.
